edx_toggles/toggles/internal/waffle.py

```diff
@@ -0,0 +1,339 @@
+for temporarily instrumenting/monitoring waffle flag usage.
+"""
+import logging
+from contextlib import contextmanager
+from weakref import WeakSet
+
+import crum
+from django.conf import settings
+from edx_django_utils.cache import RequestCache
+from edx_django_utils.monitoring import set_custom_attribute
+from waffle import flag_is_active, switch_is_active
+
+from .base import BaseNamespace, BaseToggle
+
+log = logging.getLogger(__name__)
+
+
+class WaffleSwitchNamespace(BaseNamespace):
+    """
+    Provides a single namespace for a set of waffle switches.
+
+    All namespaced switch values are stored in a single request cache containing
+    all switches for all namespaces.
+    """
+
+    def is_enabled(self, switch_name):
+        """
+        Returns and caches whether the given waffle switch is enabled.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        value = self._cached_switches.get(namespaced_switch_name)
+        if value is None:
+            value = switch_is_active(namespaced_switch_name)
+            self._cached_switches[namespaced_switch_name] = value
+        return value
+
+    @property
+    def _cached_switches(self):
+        """
+        Return a dictionary of all namespaced switches in the request cache.
+        """
+        return _get_waffle_namespace_request_cache().setdefault("switches", {})
+
+
+class WaffleSwitch(BaseToggle):
+    """
+    Represents a single waffle switch, using a cached namespace.
+    """
+
+    NAMESPACE_CLASS = WaffleSwitchNamespace
+    _class_instances = WeakSet()
+
+    def __init__(self, waffle_namespace, switch_name, module_name=None):
+        """
+        Arguments:
+            waffle_namespace (Namespace | String): Namespace for this switch.
+            switch_name (String): The name of the switch (without namespacing).
+            module_name (String): The name of the module where the flag is created. This should be ``__name__`` in most
+            cases.
+        """
+        if isinstance(waffle_namespace, str):
+            waffle_namespace = self.NAMESPACE_CLASS(name=waffle_namespace)
+
+        self.waffle_namespace = waffle_namespace
+        self.switch_name = switch_name
+
+        # Note that the waffle constructor does not provide a default
+        name = self.waffle_namespace._namespaced_name(self.switch_name)
+        super().__init__(name, default=False, module_name=module_name)
+
+    @property
+    def namespaced_switch_name(self):
+        """
+        For backward compatibility, we still provide the `namespaced_switch_name`, property, even though users should
+        now use the `name` attribute.
+        """
+        return self.name
+
+    def is_enabled(self):
+        return self.waffle_namespace.is_enabled(self.switch_name)
+
+    @contextmanager
+    def override(self, active=True):
+        with self.waffle_namespace.override(self.switch_name, active):
+            yield
```

Comment:
Why were we able to lose this for WaffleFlag, but not WaffleSwitch?  Is that just a miss, or was this intentionally left here?

Reply:
This is an oversight.